agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,315 @@
+    stats.collisions_mean /= stats.buckets_with_collisions;
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+
+  return stats;
+}
+
+static bool nr_zf_is_unnamed_closure(const zend_function* zf) {
+  if (9 != ZSTR_LEN(zf->op_array.function_name)) {
+    return false;
+  }
+  return 0 == memcmp("{closure}", ZSTR_VAL(zf->op_array.function_name), 9);
+}
+
+void nr_php_wraprec_hashmap_key_set(nr_php_wraprec_hashmap_key_t *key, const zend_function *zf) {
+  key->lineno = nr_php_zend_function_lineno(zf);
+  key->scope_name = NULL;
+  key->function_name = NULL;
+  key->filename = NULL;
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+      key->function_name = zf->op_array.function_name;
+      zend_string_addref(key->function_name);
+      if (zf->op_array.scope) {
+        key->scope_name = zf->op_array.scope->name;
+        zend_string_addref(key->scope_name);
+      }
+  } else {
+    if (NULL != zf->op_array.filename) {
+      key->filename = zf->op_array.filename;
+      zend_string_addref(key->filename);
+    }
+  }
+}
+
+void nr_php_wraprec_hashmap_key_release(nr_php_wraprec_hashmap_key_t *key) {
+  if (NULL != key->scope_name) {
+    zend_string_release(key->scope_name);
+  }
+  if (NULL != key->function_name) {
+    zend_string_release(key->function_name);
+  }
+  if (NULL != key->filename) {
+    zend_string_release(key->filename);
+  }
+  key->lineno = 0;
+}
+
+static inline size_t nr_zendfunc2bucketidx(size_t log2_num_buckets, zend_function* zf) {
+  /* default to lineno */
+  uint32_t hash = nr_php_zend_function_lineno(zf);
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    /* but use hash of function name when possible */
+    hash = ZSTR_HASH(zf->op_array.function_name);
+  } else if (NULL != zf->op_array.filename) {
+    /* and as last resort try hash of filename if available */
+    hash = ZSTR_HASH(zf->op_array.filename);
+  }
+
+  /* normalize to stay in-bound */
+  return (hash & ((1 << log2_num_buckets) - 1));
+}
+
+static inline bool zstr_equal(zend_string* zs1, zend_string* zs2) {
+  if (NULL == zs1 || NULL == zs2) {
+    return false;
+  }
+
+  if (ZSTR_LEN(zs1) != ZSTR_LEN(zs2)) {
+    return false;
+  }
+
+  if (ZSTR_HASH(zs1) != ZSTR_HASH(zs2)) {
+    return false;
+  }
+
+  if (0 != memcmp(ZSTR_VAL(zs1), ZSTR_VAL(zs2), ZSTR_LEN(zs1))) {
+    return false;
+  }
+
+  return true;
+}
+
+static bool nr_is_wraprec_for_zend_func(nr_php_wraprec_hashmap_key_t* key, zend_function* zf) {
+
+  if (nr_php_zend_function_lineno(zf) != key->lineno) {
+    return false;
+  }
+
+  if (zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    if (!zstr_equal(key->function_name, zf->op_array.function_name)) {
+      return false;
+    }
+    /* compare scope if it is set */
+    if (zf->op_array.scope && !zstr_equal(key->scope_name, zf->op_array.scope->name)) {
+      return false;
+    }
+    return true;
```

Comment:
Some comments explaining this if-else structure would be helpful

Reply:
@ZNeumann Is this what you hoped to see: d51b4ed34fce72f93bec321480d2dc2622eedbc5?